bot/exts/valentines/markov_poem_generator.py

```diff
@@ -0,0 +1,317 @@
+
+        return cache[word]
+
+    return wrapper
+
+
+class MarkovPoemGenerator(commands.Cog):
+    """
+    A cog that provides a poem by taking the markov of a corpus.
+
+    By processing the corpus text through a markov chain, a series of lines
+    can be iterated through, whilst corresponding to the given rhyme scheme.
+    """
+
+    POEM_TIMEOUT = 20  # In seconds
+
+    SOURCES: List[str] = [
+        "shakespeare_corpus.txt"
+    ]
+
+    templates: Dict[str, str] = {
+        "shakespearean-sonnet": "abab/cdcd/efef/gg",
+        "quads": "aaaa/bbbb/cccc/dddd/eeee"  # For testing
+    }
+
+    rhyme_websites: List[Tuple[bool, str]] = [
+        # (is exact rhyme, website link)
+        (True, "https://api.datamuse.com/words?rel_rhy="),
+        (False, "https://api.datamuse.com/words?rel_nry=")
+    ]
+
+    def __init__(self, bot: commands.Bot):
+        """Initializes the full corpus text and the markov model."""
+        self.bot = bot
+
+        # Load the full text corpus
+        for source_file in self.SOURCES:
+            with Path(f"bot/resources/valentines/{source_file}").open() as f:
+                full_corpus = f.read().splitlines()
+
+        # Create the markov model
+        self.model = markovify.Text(full_corpus, state_size=1)
+
+        logging.info("Full text corpus and markov model successfully loaded.")
+
+    @staticmethod
+    def _get_last_word(sentence: str) -> str:
+        """Returns the last word of a sentence string."""
+        if isinstance(sentence, str) is False:
+            # Likely to be caused by `make_short_sentence` running out
+            # and feeding it None, which is an unlikely scenario
+            raise TypeError(f"Argument is a {type(sentence)} instead of a str.")
+        return sentence.strip(string.punctuation).split()[-1]
+
+    @memoize
+    async def _get_rhyme_set(
+        self,
+        word: str,
+        near_rhyme_min_score: int = 2000
+    ) -> None:
+        """
+        Accesses web APIs to get rhymes and returns a set.
+
+        `near_rhyme_min_score` is to filter out near rhymes that barely rhyme
+        with the word. The equivalent for exact rhymes is not necessary as they
+        already rhyme.
+
+        Should additional web APIs for rhymes be added, the `min_score` needs
+        to be tuned. Perhaps it should be added as an element to the tuple of
+        the element of the `self.rhyme_websites` list.
+        """
+        rhyme_set = set()
+
+        for is_exact, website in self.rhyme_websites:
+            min_score = 0 if is_exact else near_rhyme_min_score
+
+            async with self.bot.http_session.get(
+                website + word,
+                timeout=10
+            ) as response:
+                curr_set = set(
+                    data["word"] for data in await response.json()
+                    if data.get("score", 0) >= min_score
+                )
+                rhyme_set |= curr_set
+
```

Comment:
This part should only be done if getting the information from the api was successful i.e. the status code of the response was `200`.
```suggestion
                if response.status == 200:
                    curr_set = set(
                        data["word"] for data in await response.json()
                        if data.get("score", 0) >= min_score
                    )
                    rhyme_set |= curr_set
                else:
                    logging.warning(f"Received response {response.status} from: {website + word}")

```